ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -382,6 +445,17 @@ protected String getUuid() {
         idBuf.putShort(accountId);
         idBuf.putShort(containerId);
         break;
+      case BLOB_ID_V5:
+        flag = (byte) (type.ordinal() & BLOB_ID_TYPE_MASK);
+        flag |= isEncrypted ? IS_ENCRYPTED_MASK : 0;
+        if (blobDataType != null) {
```

Comment:
why would it be `null` in version 5?

Reply:
It shouldn't, but I was getting NPE due a test not setting.  That test may be fixed now but it doesn't hurt to check for null.

Reply:
My preference is don't add extra check for `blobDataType `:
If not because of tests, might be a corner case that is not covered. Throwing NPE is helpful to debug. The similar variable is `isEncrypted `, no extra check is needed.
If it's because of tests, then must be something wrong in test. Let's fix it.

Reply:
+1 - no null checks here. It shouldn't _be_ `null` (i.e `blobDataType == null` is an `Exception` scenario)

Reply:
Seems like doing this would cause you to change a lot of the tests - have to chat about how to make this work

Reply:
@vgkholla @zzmao I tried enforcing non-null, but this fails in the blobid crafting scenario which takes an existing blobId with version < 5 and up-versions it to the latest (5).  The original blobId has null data type.  How do you propose to handle this case?  I think we need to allow null value, and @cgtz seems to agree.

Reply:
If we allow null value, it may mask future bug. 
We can modify the test to do sth special for version<5.

Reply:
@zzmao It's not the test, it's the crafting functionality itself.  What do you suggest?

Reply:
I don't think we need to worry about crafting too much since it is an offline operation performed with a specific goal in mind - which means that it is reasonable to expect that the crafting script has all the info required to construct a blob ID in the version required.

Reply:
@vgkholla but are they/we able to know `blobDataType` or `metaDataType`(v1-v4) before run crafting script? What if unknown?

Reply:
then a v5 version of v1-v4 cannot be created. The blob must also be crafted in the same version.

Reply:
Ah. I see. Then we are safe.